README.md
# meetup-web-mocks

[![npm version](https://badge.fury.io/js/meetup-web-mocks.svg)](https://badge.fury.io/js/meetup-web-mocks)
[![Build Status](https://travis-ci.org/meetup/meetup-web-mocks.svg?branch=master)](https://travis-ci.org/meetup/meetup-web-components)

Mock data for use across meetup web repositories such as `meetup-web-platform`, `meetup-web-components`, `mup-web`, etc.

Data includes mocked api endpoints and mocked app data such as requests, state, etc.


CHANGELOG.md
# [1.2.x]

- Migrate the build to Github Action

# [1.1.x]

- Upgrade to node v12


src/api.js
export const MOCK_MEMBER = {
	name: 'mocky mcmockerson',
	id: 1243,
	photo: {
		photo_link: 'http://placekitten.com/g/400/400',
		thumb_link: 'http://placekitten.com/g/200/200',
	},
};

export const MOCK_SELF = { ...MOCK_MEMBER, ...{ lang: 'en_US' } };
export const MOCK_SELF_FR = { ...MOCK_MEMBER, ...{ lang: 'fr_FR' } };

export const MOCK_DUOTONES = [['a', 'b']];
export const MOCK_DUOTONE_URLS = {
	dtaxb: {
		small: 'http://a.b',
		large: 'http://a.b',
	},
};

export const MOCK_GROUP = {
	id: 1234,
	urlname: 'fake-hq-faff',
	organizer: {
		id: 164510652,
		name: 'Grayson Cheese',
		bio:
			'Org bio lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent eu tellus sit amet erat blandit volutpat.',
		photo: {
			id: 257204155,
			highres_link:
				'http://photos1.dev.meetupstatic.com/photos/member/a/c/7/b/highres_257204155.jpeg',
			photo_link:
				'http://photos1.dev.meetupstatic.com/photos/member/a/c/7/b/member_257204155.jpeg',
			thumb_link:
				'http://photos1.dev.meetupstatic.com/photos/member/a/c/7/b/thumb_257204155.jpeg',
			type: 'member',
			base_url: 'http://photos1.dev.meetupstatic.com',
		},
	},
	plain_text_description:
		'Plain text description lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent eu tellus sit amet erat blandit volutpat. Suspendisse tincidunt pulvinar vehicula. Mauris eget tristique mi. Ut tincidunt, ante eget aliquet eleifend, ligula lorem elementum felis, quis consectetur sapien justo ut eros. Maecenas tempus lacus metus, eu pretium massa convallis vitae. Donec id fringilla orci. Nunc eget elementum massa. Pellentesque pulvinar ipsum porta finibus laoreet. Donec pharetra quis nunc in auctor. Interdum et malesuada fames ac ante ipsum primis in faucibus. Suspendisse potenti. Nulla eget nibh quis nisi commodo accumsan. Donec sed suscipit justo. Aliquam erat volutpat. Nunc a nulla massa. Curabitur at elit eu neque congue fermentum',
	key_photo: {
		id: 1234,
		photo_link: 'http://placekitten.com/400/300',
		thumb_link: 'http://placekitten.com/400/300',
	},
	group_photo: {
		id: 1234,
		photo_link: 'http://placekitten.com/400/300',
		thumb_link: 'http://placekitten.com/400/300',
	},
	name: 'fake HQ FAFF',
	link: 'http://example.com/fake-hq-faff',
	who: 'Faffers',
	members: 999,
	photo_gradient: {
		light_color: MOCK_DUOTONES[0][1],
		dark_color: MOCK_DUOTONES[0][0],
	},
	duotoneUrl: MOCK_DUOTONE_URLS.dtaxb,
	event_sample: [
		{
			id: 1234,
			name: 'Sample event',
			time: 1515625071128,
			yes_rsvp_count: 50,
		},
	],
	timezone: 'Europe/London',
};

/* Updated shape for members list re-platform */
export const MOCK_MEMBER_LIST_MEMBER = {
	name: 'Thomas Shelby',
	last_visited: '2018-10-12T15:15:35.000Z',
	joined: '2017-03-20T11:07:29.000Z',
	id: '1',
	self: {
		actions: ['message'],
	},
	status: 'active',
	photo: {
		id: '255539096',
		thumb_link:
			'https://static.spin.com/files/2016/07/cillian-murphy-peaky-blinders-loves-radiohead-interview-640x426.jpg',
	},
	role: 'coorganizer',
	dues: {
		status: 'paid',
		paid_end_date: '2020-03-20T11:07:29.000Z',
	},
};

export const MOCK_MEMBER_LIST_MEMBER_2 = {
	...MOCK_MEMBER_LIST_MEMBER,
	name: 'Arthur Shelby',
	last_visited: '2017-10-12T15:15:35.000Z',
	joined: '2016-03-20T11:07:29.000Z',
	id: '12',
	photo: {
		id: '458585',
		thumb_link: 'https://ichef.bbci.co.uk/images/ic/480xn/p05mk5c6.jpg',
	},
	dues: {
		status: 'exempt',
	},
	role: null,
};

export const MOCK_MEMBER_LIST_MEMBER_3 = {
	...MOCK_MEMBER_LIST_MEMBER,
	name: 'John Shelby',
	last_visited: '2017-12-12T15:15:35.000Z',
	joined: '2016-06-20T11:07:29.000Z',
	id: '123',
	photo: {
		id: '992929',
		thumb_link:
			'https://vignette.wikia.nocookie.net/peaky-blinders/images/f/f0/John2.jpg/revision/latest?cb=20171102225430',
	},
	dues: {
		status: 'trial',
		trial_end_date: '2019-03-20T11:07:29.000Z',
	},
	role: null,
};

export const MOCK_MEMBER_LIST_MEMBER_4 = {
	...MOCK_MEMBER_LIST_MEMBER,
	name: 'Polly Grey',
	last_visited: '2017-12-12T15:15:35.000Z',
	joined: '2016-06-20T11:07:29.000Z',
	id: '1234',
	photo: {
		id: '12222',
		thumb_link: 'https://ichef.bbci.co.uk/images/ic/480xn/p01g4m2n.jpg',
	},
	dues: {
		status: 'unpaid',
	},
	role: 'event_organizer',
};

export const MOCK_MEMBER_LIST_MEMBER_5 = {
	...MOCK_MEMBER_LIST_MEMBER,
	name: 'Alfie Solomons',
	last_visited: null,
	joined: null,
	id: '12345',
	photo: {
		id: '124124',
		thumb_link:
			'https://i.pinimg.com/originals/e7/70/73/e770734b59f8184f08565214c170bf53.jpg',
	},
	role: null,
	dues: null,
	status: 'banned',
};

export const MOCK_MEMBER_LIST_MEMBER_6 = {
	...MOCK_MEMBER_LIST_MEMBER,
	name: 'Ada Shelby',
	last_visited: null,
	joined: null,
	id: '123456',
	photo: {
		id: '124124',
		thumb_link: 'https://ichef.bbci.co.uk/images/ic/480xn/p05mk5w9.jpg',
	},
	role: null,
	dues: null,
	status: 'pending',
};

export const MOCK_MEMBERS_LIST_ALL = [
	MOCK_MEMBER_LIST_MEMBER,
	MOCK_MEMBER_LIST_MEMBER_2,
	MOCK_MEMBER_LIST_MEMBER_3,
	MOCK_MEMBER_LIST_MEMBER_4,
];

export const MOCK_MEMBERS_LIST_LEADERSHIP = [
	MOCK_MEMBER_LIST_MEMBER,
	MOCK_MEMBER_LIST_MEMBER_4,
];

export const MOCK_MEMBERS_LIST_PENDING = [MOCK_MEMBER_LIST_MEMBER_6];

export const MOCK_MEMBERS_LIST_BANNED = [MOCK_MEMBER_LIST_MEMBER_5];

export function getMockMemberList(selectedFilter) {
	if (selectedFilter === 'all') {
		return MOCK_MEMBERS_LIST_ALL;
	}
	if (selectedFilter === 'leaders') {
		return MOCK_MEMBERS_LIST_LEADERSHIP;
	}
	if (selectedFilter === 'pending') {
		return MOCK_MEMBERS_LIST_PENDING;
	}
	return MOCK_MEMBERS_LIST_BANNED;
}

export const MOCK_CHAPTER_MEMBER = {
	id: 82159492,
	name: 'Grayson Cheese',
	status: 'active',
	joined: 1361806241000,
	city: 'new york',
	country: 'us',
	lat: 40,
	lon: -72,

	photo: {
		photo_link: 'http://placekitten.com/g/400/400',
	},
	group_profile: {
		status: 'active',
		visited: 1396285003000,
		created: 1395768492000,
		updated: 1478880527000,
		group: MOCK_GROUP,
		intro: 'cheese is the best!',
	},
};
export const MOCK_CHAPTER_MEMBER_2 = {
	id: 234322,
	name: 'Chesterton Gordidason',
	status: 'active',
	joined: 1361806241020,
	city: 'new york',
	country: 'us',
	lat: 40,
	lon: -72,

	photo: {
		photo_link: 'http://placekitten.com/g/400/400',
	},
	group_profile: {
		status: 'active',
		visited: 1396285003000,
		created: 1395768492000,
		updated: 1478880527000,
		group: MOCK_GROUP,
		intro: 'cheese is the best!',
	},
};
export const MOCK_CHAPTER_MEMBER_3 = {
	id: 23222,
	name: 'Simon Cheddar',
	status: 'active',
	joined: 1361806541020,
	city: 'new york',
	country: 'us',
	lat: 40,
	lon: -72,

	photo: {
		photo_link: 'http://placekitten.com/g/400/400',
	},
	group_profile: {
		status: 'active',
		visited: 1396285003000,
		created: 1395768492000,
		updated: 1478880527000,
		group: MOCK_GROUP,
		intro: 'cheese is the best!',
	},
};

export const MOCK_VENUE = {
	address_1: 'address 1',
	address_2: 'address 2',
	address_3: 'address 3',
	city: 'city name',
	state: 'NY',
	zip: '00000',
};

export const MOCK_COMMENT = {
	id: '1234',
	comment: 'comment 1',
	created: 1479838415,
	like_count: 6,
	member: MOCK_CHAPTER_MEMBER,
	deleted: false,
	visible: true,
	updated: 1479838415,
	utc_offset: -18000000,
	self: {
		liked: false,
		actions: ['like', 'delete', 'flag_spam'],
	},
};

export const MOCK_COMMENTS_LIST = {
	0: MOCK_COMMENT,
	1: {
		...MOCK_COMMENT,
		id: '12345',
		comment:
			'comment 2, Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam tincidunt facilisis dui ac tincidunt. Duis rutrum bibendum odio nec tincidunt. Suspendisse sagittis blandit venenatis.',
		created: 1478879940,
		like_count: 3,
		self: {
			liked: false,
			actions: ['like', 'delete', 'flag_spam'],
		},
	},
	2: {
		...MOCK_COMMENT,
		id: 21234,
		comment:
			'Morbi vel purus eu neque ultricies maximus. Nunc sollicitudin maximus justo sed fringilla.',
		created: 1478879939,
		like_count: 22,
		self: {
			liked: false,
			actions: ['like', 'flag_spam'],
		},
	},
	3: {
		...MOCK_COMMENT,
		id: 23423,
		comment:
			'Sed odio tortor, vulputate ut tincidunt vitae, dignissim eget felis.',
		created: 1478879939,
		like_count: 0,
		self: {
			liked: false,
			actions: ['like', 'delete', 'flag_spam'],
		},
	},
	4: {
		...MOCK_COMMENT,
		id: 234,
		comment:
			'Curabitur elit nisi, porta et elit non, efficitur sodales augue. Phasellus volutpat, lectus ac tempor laoreet, diam urna commodo lacus, eget bibendum risus augue a sem.',
		created: 1478879939,
		like_count: 0,
		self: {
			liked: false,
			actions: ['like', 'delete', 'flag_spam'],
		},
	},
};

const oneMonthAgo = new Date();
oneMonthAgo.setMonth(oneMonthAgo.getMonth() - 1);
const oneMonthFuture = new Date();
oneMonthFuture.setMonth(oneMonthFuture.getMonth() + 1);
export const MOCK_EVENT = {
	id: '123456',
	comment_count: 5,
	created: oneMonthAgo.getTime(),
	description: `The coolest event in the world during which we will run and dance and sing
	<script>alert("bad time")</script>, 😊, &lt;blink&gt;what what&lt;blink&gt; this is getting
	longer than it needs to be why am I still typing omg`,
	duration: 3600000,
	name: 'So much fun',
	rsvp_sample: [
		{
			created: 1462833255609,
			id: 1234,
			member: MOCK_MEMBER,
			updated: 1462833255610,
		},
		{
			created: 1462833255609,
			id: 2345,
			member: { ...MOCK_MEMBER, ...{ id: 8912894 } },
			updated: 1462833255610,
		},
		{
			created: 1462833255609,
			id: 3456,
			member: { ...MOCK_MEMBER, ...{ id: 899828 } },
			updated: 1462833255610,
		},
	],
	venue: MOCK_VENUE,
	how_to_find_us: 'how to find us',
	event_hosts: [
		MOCK_CHAPTER_MEMBER,
		MOCK_CHAPTER_MEMBER_2,
		MOCK_CHAPTER_MEMBER_3,
	],
	fee: {
		amount: 10,
		currency: 'USD',
	},
	rsvpable: true,
	group: MOCK_GROUP,
	self: {
		actions: ['rsvp'],
		pay_status: 'none',
		rsvp: {},
	},
	status: 'upcoming',
	time: oneMonthFuture.getTime(),
	utc_offset: 0,
	visibility: 'public',
	yes_rsvp_count: 23,
	link: 'https://www.meetup.com/group/events/123456',
};

// Mock category sourced from https://www.meetup.com/meetup_api/docs/find/topic_categories/
export const MOCK_CATEGORY = {
	id: 552,
	shortname: 'movements',
	name: 'Movements',
	sort_name: 'Movements',
	photo: {
		id: 450131941,
		highres_link:
			'http://photos1.meetupstatic.com/photos/event/2/e/a/5/highres_450131941.jpeg',
		photo_link:
			'http://photos1.meetupstatic.com/photos/event/2/e/a/5/600_450131941.jpeg',
		thumb_link:
			'http://photos1.meetupstatic.com/photos/event/2/e/a/5/thumb_450131941.jpeg',
		type: 'event',
		base_url: 'http://photos1.meetupstatic.com',
	},
	category_ids: [4, 13],
	best_topics: [
		{
			id: 16645,
			name: 'Volunteering',
			urlkey: 'volunteering',
			lang: 'en_US',
		},
	],
};

export const MOCK_CATEGORIES = [MOCK_CATEGORY];

export const MOCK_EVENT_SCHEDULE_RESPONSE = {
	ref: 'eventSchedule',
	value: {
		created: 1505337385000,
		duration: 7200000,
		id: 'cgpwfnywmbsb',
		name: 'A Weekly Hike',
		status: 'upcoming',
		time: 1505430000000,
		updated: 1505337385000,
		utc_offset: -14400000,
		waitlist_count: 0,
		yes_rsvp_count: 0,
		group: {
			created: 1401198587000,
			name: 'Managing Meetup',
			id: 14705172,
			join_mod: 'approval',
			lat: 40.720001220703125,
			lon: -74,
			urlname: 'ManagingMeetup',
			who: 'Managers',
			localized_location: 'New York, NY',
			region: 'en_US',
		},
		link: 'https://www.dev.meetup.com/ManagingMeetup/events/cgpwfnywmbsb/',
		description:
			"<p>• What we'll do<br/>Hiking often! Mountains Forests all types of trails</p> <p>• What to bring<br/>Shoes and water</p>",
		visibility: 'public_limited',
	},
	meta: {
		server: 'ivy',
		requestId: 'c0ae499a-a473-43c5-94f8-f457c66f6c05',
		endpoint: 'ManagingMeetup/events/',
		statusCode: 201,
	},
};

export const MOCK_EVENT_SCHEDULE_RESPONSE_ERRORS = {
	ref: 'eventSchedule',
	value: {
		errors: [
			{
				code: 'name_error',
				message: 'A title is required to post your Meetup',
				field: 'name',
			},
		],
	},
	error: 'Bad Request',
	meta: {
		server: 'ivy',
		requestId: 'a6d2f9d3-9c98-416e-8f51-5753384c8de2',
		endpoint: 'ManagingMeetup/events/',
		statusCode: 400,
	},
};

/*
 * Shallow mocks that don't include dynamic dates
 * for ease of use with snapshots.
 */

export const MOCK_SHALLOW_GROUP = {
	id: MOCK_GROUP.id,
	key_photo: MOCK_GROUP.key_photo,
	name: MOCK_GROUP.name,
	urlname: MOCK_GROUP.urlname,
	members: MOCK_GROUP.members,
	organizer: MOCK_GROUP.organizer,
	link: MOCK_GROUP.link,
	duotoneUrl: MOCK_GROUP.duotoneUrl,
	group_photo: MOCK_GROUP.group_photo,
	photo_gradient: MOCK_GROUP.photo_gradient,
	plain_text_description: MOCK_GROUP.plain_text_description,
	who: MOCK_GROUP.who,
	timezone: MOCK_GROUP.timezone,
};
export const MOCK_TIME = 1508279299118;
export const MOCK_SHALLOW_MEMBER = { id: MOCK_CHAPTER_MEMBER.id };
export const MOCK_SHALLOW_EVENT = {
	id: '123456',
	comment_count: 5,
	created: MOCK_TIME,
	description:
		'The coolest event in the world during which we will run and dance and sing\n\t<script>alert("bad time")</script>, \uD83D\uDE0A, &lt;blink&gt;what what&lt;blink&gt; this is getting\n\tlonger than it needs to be why am I still typing omg',
	duration: 3600000,
	name: 'So much fun',
	rsvp_sample: [
		{
			created: 1462833255609,
			id: 1234,
			member: MOCK_SHALLOW_MEMBER,
			updated: 1462833255610,
		},
		{
			created: 1462833255609,
			id: 2345,
			member: { ...MOCK_SHALLOW_MEMBER, id: 8912894 },
			updated: 1462833255610,
		},
		{
			created: 1462833255609,
			id: 3456,
			member: { ...MOCK_SHALLOW_MEMBER, id: 899828 },
			updated: 1462833255610,
		},
	],
	venue: MOCK_VENUE,
	how_to_find_us: 'how to find us',
	event_hosts: [],
	fee: {
		amount: 10,
		currency: 'USD',
	},
	rsvpable: true,
	group: MOCK_SHALLOW_GROUP,
	self: {
		actions: ['rsvp'],
		pay_status: 'none',
		rsvp: {},
	},
	status: 'upcoming',
	time: MOCK_TIME,
	utc_offset: 0,
	visibility: 'public',
	yes_rsvp_count: 23,
};

export const MOCK_EVENTS = [
	{ ...MOCK_SHALLOW_EVENT, id: '1111', status: 'upcoming' },
	{ ...MOCK_SHALLOW_EVENT, id: '2222', status: 'upcoming' },
	{ ...MOCK_SHALLOW_EVENT, id: '3333', status: 'upcoming' },
	{ ...MOCK_SHALLOW_EVENT, id: '4444', status: 'upcoming' },
];

export const MOCK_PAST_EVENTS = [
	{ ...MOCK_SHALLOW_EVENT, id: '5555', status: 'past' },
	{ ...MOCK_SHALLOW_EVENT, id: '6666', status: 'past' },
	{ ...MOCK_SHALLOW_EVENT, id: '7777', status: 'past' },
	{ ...MOCK_SHALLOW_EVENT, id: '8888', status: 'past' },
];

export const UPCOMING_EVENT = { ...MOCK_EVENTS[0], status: 'upcoming' };
export const PROPOSED_EVENT = { ...MOCK_EVENTS[1], status: 'proposed' };
export const PAST_EVENT = { ...MOCK_EVENTS[2], status: 'past' };

export const MOCK_FEEDBACK_SUMMARY_RESPONSE = {
	eventId: '241470859',
	summary: [
		{
			questionId: 'wouldYouGo',
			answers: [
				{
					answerId: 'wouldYouGoYes',
					value: 5,
				},
				{
					answerId: 'wouldYouGoNo',
					value: 3,
				},
			],
		},
		{
			questionId: 'whatsWorthwhile',
			answers: [
				{
					answerId: 'people',
					value: 3,
				},
				{
					answerId: 'purpose',
					value: 2,
				},
			],
		},
		{
			questionId: 'whyNot',
			answers: [
				{
					answerId: 'hardToFind',
					value: 2,
				},
				{
					answerId: 'badVenue',
					value: 1,
				},
			],
		},
	],
};


.eslintrc.js
const config = {
	extends: 'eslint:recommended',
	parserOptions: {
		ecmaVersion: 6,
		ecmaFeatures: {
			jsx: true,
			experimentalObjectRestSpread: true,
		},
		sourceType: 'module',
	},
	env: {
		browser: true,
		node: true,
		jest: true,
		jasmine: true,
		es6: true,
	},
	plugins: ['react'],
	rules: {
		'array-callback-return': 2,
		'comma-dangle': 0,
		'eol-last': 2,
		indent: [2, 'tab'],
		'jsx-quotes': [1, 'prefer-single'],
		'no-console': 0,
		'no-empty-pattern': 2,
		'no-lone-blocks': 2,
		'no-lonely-if': 2,
		'no-multi-spaces': 2,
		'no-negated-condition': 2,
		'no-trailing-spaces': 2,
		'no-unused-vars': [2, { args: 'none' }],
		'no-use-before-define': 2,
		'no-useless-call': 2,
		'no-useless-concat': 2,
		'no-useless-escape': 2,
		'no-var': 2,
		'no-whitespace-before-property': 2,
		'prefer-const': 2,
		'prefer-template': 2,
		'quote-props': [2, 'as-needed'],
		'require-jsdoc': [
			2,
			{
				require: {
					FunctionDeclaration: false,
					ClassDeclaration: true,
				},
			},
		],
		'react/jsx-no-undef': 0,
		'react/jsx-uses-react': 1,
		'react/jsx-uses-vars': 1,
		'react/no-deprecated': 2,
		'react/no-string-refs': 1,
		'react/react-in-jsx-scope': 2,
		'react/jsx-wrap-multilines': 2,
		'react/jsx-boolean-value': 2,
		'react/jsx-equals-spacing': [2, 'never'],
		'react/jsx-no-duplicate-props': 2,
		'react/jsx-pascal-case': 2,
		'react/jsx-uses-react': 2,
		semi: 2,
		'space-in-parens': [2, 'never'],
		'spaced-comment': [2, 'always'],
		strict: 0,
	},
};

module.exports = config;


src/app.js
import {
	MOCK_GROUP,
	MOCK_EVENT,
	MOCK_SHALLOW_GROUP,
	MOCK_SHALLOW_EVENT,
} from './api';
import { MOCK_COMM_APP_STATE } from './group/communication/app';

const api = {
	self: {
		type: 'member',
		value: {
			id: 1234,
			status: 'prereg',
		},
	},
	group: {
		type: 'group',
		value: MOCK_GROUP,
	},
	events: {
		type: 'event',
		value: [
			{ ...MOCK_EVENT, ...{ id: '3456' } },
			{ ...MOCK_EVENT, ...{ id: '4567' } },
		],
	},
	event: {
		type: 'event',
		value: MOCK_EVENT,
	},
	[`event_${MOCK_SHALLOW_GROUP.urlname}_${MOCK_SHALLOW_EVENT.id}`]: {
		type: 'event',
		value: MOCK_SHALLOW_EVENT,
	},

	...MOCK_COMM_APP_STATE,
};

export const MOCK_APP_STATE = {
	app: api,
	api,
	config: {
		apiUrl: 'http://fake.api.meetup.com',
		csrf: '1234_fake_csrf',
		requestLanguage: 'en-US',
		baseUrl: 'https://beta2.dev.meetup.com:8000',
	},
	routing: {
		locationBeforeTransitions: {
			pathname: '/foo',
		},
	},
	toasts: { ready: [] },
};

export const MOCK_API_PROBLEM = {
	problem: 'There was an internal problem in the API',
};

export const MOCK_API_RESULT = [
	{
		ref: 'group',
		type: 'group',
		value: MOCK_APP_STATE.app.group.value,
	},
];

export const MOCK_FETCH_API_RESPONSE = {
	queries: [],
	responses: [],
};

export const MOCK_OAUTH_COOKIES = {
	oauth_token: '1234',
	refresh_token: 'asdf',
	anonymous: true,
};

export const MOCK_RENDERPROPS = {
	location: {
		// https://github.com/reactjs/history/blob/master/docs/Location.md
		pathname: '/foo',
		search: '?foo',
		state: {},
		action: 'PUSH',
		key: '1234',
		query: {
			foo: null,
		},
	},
	params: {
		urlname: 'foo',
	},
};

export const MOCK_RENDERPROPS_UTF8 = {
	location: {
		// https://github.com/reactjs/history/blob/master/docs/Location.md
		pathname: '/バ-京',
		search: '',
		state: {},
		action: 'PUSH',
		key: '1234',
	},
	params: {
		urlname: 'バ-京',
	},
};

export const MOCK_MEANINGLESS_ACTION = {
	type: 'ARBITRARY',
	payload: '/',
};

export const mockQuery = ({ location, params, apiMetaHeaders }) => {
	const meta = {};
	if (apiMetaHeaders) {
		meta.apiMetaHeaders = apiMetaHeaders;
	}
	return {
		type: 'group',
		endpoint: (params || {}).urlname || 'foo',
		params: params,
		meta,
		ref: 'group',
	};
};

export const mockQueryBadType = ({ location, params }) => {
	const type = 'lkajlhsdhljaskliub';
	return { ...mockQuery({ location, params }), type };
};

export const MOCK_AUTH_HEADER = `Bearer ${MOCK_OAUTH_COOKIES}.oauth_token`;
export const MOCK_DATETIME = new Date().getTime();
export const MOCK_CREDENTIALS = {
	username: 'Michael McGahan',
	pw: 'whatever',
};

export const MOCK_LOGIN_RESPONSE = {
	value: {
		member: {},
		oauth_token: 1234,
	},
};

export const MOCK_POST_ACTION = {
	type: 'POST_DUMMY',
	payload: {
		query: {
			endpoint: 'mypost',
			ref: 'postiepost',
			type: 'mcposterson',
		},
		onSuccess: payload => ({ type: 'MOCK_SUCCESS', payload }),
		onError: err => ({ type: 'MOCK_ERROR', payload: err, error: true }),
	},
};

export const MOCK_DELETE_ACTION = {
	type: 'DELETE_DUMMY',
	payload: {
		query: {
			endpoint: 'mydelete',
			ref: 'deleted',
			type: 'dellllleeeeted',
		},
		onSuccess: payload => ({ type: 'MOCK_SUCCESS', payload }),
		onError: err => ({ type: 'MOCK_ERROR', payload: err, error: true }),
	},
};

// Arbitrary string response payload from server render
export const MOCK_RENDER_RESULT =
	'<html><body><h1>Hello world</h1></body></html>';

// string 'Cookie:' header from MOCK_OAUTH_COOKIES
export const MOCK_COOKIE_HEADER = Object.keys(MOCK_OAUTH_COOKIES).reduce(
	(acc, key) => (acc += `${key}=${JSON.stringify(MOCK_OAUTH_COOKIES[key])}; `),
	''
);

export const MOCK_REQUEST_COOKIES = {
	url: '/',
	headers: {
		cookie: MOCK_COOKIE_HEADER,
	},
};

const mock_valid_config = {
	DEV_SERVER_PORT: 8000,
	API_PROTOCOL: 'https',
	API_HOST: 'api.dev.meetup.com',
	OAUTH_AUTH_URL: 'https://secure.dev.meetup.com/oauth2/authorize',
	OAUTH_ACCESS_URL: 'https://secure.dev.meetup.com/oauth2/access',
	PHOTO_SCALER_SALT: 'abcdefg',
	AUTH_ENDPOINT: '/auth',
	oauth: {
		secret: process.env.MUPWEB_OAUTH_SECRET,
		key: process.env.MUPWEB_OAUTH_KEY,
	},
};
mock_valid_config.API_SERVER_ROOT_URL = `${mock_valid_config.API_PROTOCOL}://${
	mock_valid_config.API_HOST
}`;
export const MOCK_VALID_CONFIG = mock_valid_config;

export const MOCK_ROUTES = [
	{
		path: '/noQuery',
		component: () => {},
	},
	{
		path: '/',
		component: () => {},
		query: () => mockQuery({}),
		childRoutes: [
			{
				path: 'foo',
				component: () => {},
				query: () => mockQuery({}),
			},
			{
				path: 'nullQuery',
				component: () => {},
				query: () => null,
			},
		],
	},
];

// default value object used in mock state object
export const mockValue = { id: '🦄' };
/**
 * Useful function when generating fake state objects
 * @param {String} ref Reference to use in object creation
 * @param {Object} value value to use in object creation
 * @return {Object} mock object
 */
export const generateMockState = (ref, value = mockValue) => ({
	config: {},
	api: {
		...MOCK_APP_STATE.api,
		[ref]: {
			ref,
			value,
			meta: {
				statusCode: 204,
				flags: { some_flag: true },
			},
		},
	},
});

/**
 * Useful function when generating fake state objects with a custom meta object.
 * @param {String} ref Reference to use in object creation
 * @param {Object} value value to use in object creation
 * @param {Object} meta meta values to use in object creation
 * @return {Object} mock object
 */
export const generateMockStateAndMeta = (ref, value = mockValue, meta) => ({
	config: {},
	api: {
		...MOCK_APP_STATE.api,
		[ref]: {
			ref,
			value,
			meta,
		},
	},
});

/**
 * Useful testing util when testing selectors to determine if variants
 * object will be returned for the specified ref
 * @param {String} ref Reference to use in object creation
 * @param {Object} variants
 * @param {Object} value additional api value on the response
 * @return {Object} mock object
 */
export const generateMockVariantMeta = (ref, variants, value) => {
	const mockState = generateMockState();
	return {
		api: {
			...mockState.api,
			[ref]: {
				value,
				meta: { variants },
			},
		},
	};
};
